### PR TITLE
Df field dependency

### PR DESCRIPTION
**Pull request by jbrezmorf**

### CODE REVIEW

File: src/fields/field_set.cc [hunk 1]
```diff
@@ -232,6 +232,69 @@ void FieldSet::update_coords_caches(ElementCacheMap &cache_map) {
...
 
+void FieldSet::cache_update(ElementCacheMap &cache_map) {
+    update_coords_caches(cache_map);
+    for (unsigned int i_reg=0; i_reg<mesh_->region_db().size(); ++i_reg) {
+        for(unsigned int i_f=0; i_f<region_dependency_list_[i_reg].size(); ++i_f) field_list[region_dependency_list_[i_reg][i_f]]->cache_update(cache_map, i_reg);
+    }
+}
+
+
+unsigned int FieldSet::compute_depth(const FieldCommon *field, const map<string, vector<const FieldCommon *>> &dependency_map) {
```
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/fields/field_set.cc:243]: ... unrelated, but good to do anyway.

[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/fields/field_set.cc:243]: Comment to the actual code:
- rename to dependency_dfs(...)
- use classical DFS, see e.g. https://www.tutorialspoint.com/cplusplus-program-to-apply-dfs-to-perform-the-topological-sorting-of-a-directed-acyclic-graph
- do not construct `dependency_map` before the sort, just call it before processing the input fields of a field.
- no need to compute depth, just push_back to the `field_update_order` on the postvisit. Then the order is reversed. 
- an auxiliary map or set to mark visited fields is necessary, the finishing condition is missing in this code.

File: src/fields/field_set.cc [hunk 2]
```diff
@@ -232,6 +232,69 @@ void FieldSet::update_coords_caches(ElementCacheMap &cache_map) {
...
+	        field_indices_map_[ field_list[i]->name() ] = i;
+	    }
+	}
+	region_dependency_list_.clear();
+
+    map<string, vector<const FieldCommon *>> dependency_map;
+    map<unsigned int, vector<string>> depth_map;
+    set<string> used_fields;
+	for (unsigned int i_reg=0; i_reg<mesh_->region_db().size(); ++i_reg) {
+		dependency_map["X"] = vector<const FieldCommon *>(); // Temporary solution, remove after replace coord data cache with Field
```
Comment (jbrezmorf) [inline src/fields/field_set.cc:269]: Can be done.

File: src/fields/field_set.hh [hunk 3]
```diff
@@ -296,6 +303,20 @@ protected:
     FieldValueCache<double> z_coord_;  ///< Holds values of z-coordinates (for FieldFormula)
     const Mesh *mesh_;                 ///< Pointer to the mesh.
 
+    /// Holds map of indices of fields stored in field_list
+    std::map<std::string, unsigned int> field_indices_map_;
```
Comment (jbrezmorf) [inline src/fields/field_set.hh:307]: Make a local variable of topological sort.

File: src/fields/multi_field.impl.hh [hunk 4]
```diff
@@ -377,10 +377,15 @@ bool MultiField<spacedim, Value>::MultiFieldFactory::is_active_field_descriptor(
...
 template<int spacedim, class Value>
-void MultiField<spacedim, Value>::set_dependency(FieldSet &field_set) {
+std::vector<const FieldCommon *> MultiField<spacedim, Value>::set_dependency(FieldSet &field_set, unsigned int i_reg) {
+    std::vector<const FieldCommon *> depend_fields;
     for(unsigned int i_comp=0; i_comp < this->shared_->comp_names_.size(); i_comp++) {
-        sub_fields_[i_comp].set_dependency(field_set);
+        auto add = sub_fields_[i_comp].set_dependency(field_set, i_reg);
+        depend_fields.insert(depend_fields.end(), add.begin(), add.end());
     }
+    std::sort(depend_fields.begin(), depend_fields.end());
```
Comment (jbrezmorf) [inline src/fields/multi_field.impl.hh:386]: No need to making the list unique, DFS should work even for that, as long as we have visited fields marked.

File: src/fields/field_set.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbrezmorf) [reply src/fields/field_set.cc:243]: As we sort graph with reversed edges. We use order directly produced by DFS topological sort. i.e. we push_back into a vector in the postvisit and use that vector to process fields in the correct order.

File: src/fields/field.impl.hh [hunk 5]
```diff
@@ -756,22 +756,23 @@ void Field<spacedim, Value>::cache_reallocate(const ElementCacheMap &cache_map)
...
 }
 
+
 template<int spacedim, class Value>
-void Field<spacedim, Value>::set_dependency(FieldSet &field_set) {
-    for( auto field_ptr : region_fields_) {
-    	if (field_ptr != nullptr) field_ptr->set_dependency(field_set);
-    }
+std::vector<const FieldCommon *> Field<spacedim, Value>::set_dependency(FieldSet &field_set, unsigned int i_reg) {
+   	if (region_fields_[i_reg] != nullptr) return region_fields_[i_reg]->set_dependency(field_set);
```
[the remark this answers is not in the document]
Comment (dflanderka) [reply src/fields/field.impl.hh:771]: Bounadry regions for bulk fields and vice versa are set to nullptr.

**Pull request merged**